### DISCUSSION TITLE
Missing Features Implemented by Other Text Expanders

### DISCUSSION THREAD

**Discussion by CharlesARoy** [espanso/espanso#1528] [Feature requests and Ideas]
I've used and tried a few text expanders over the years including:
[Alfred](https://www.alfredapp.com/)
[aText](https://www.trankynam.com/atext/)
[Dash](https://kapeli.com/dash)
[FastKeys](https://www.fastkeysautomation.com/)
[PhraseExpander](https://www.phraseexpander.com/)
[PhraseExpress](https://www.phraseexpress.com/)
[Rocket Typist](https://www.witt-software.com/rockettypist/)
[TextExpander](https://textexpander.com/)
[Typinator](https://ergonis.com/typinator)
[TypeIt4Me](https://apps.apple.com/us/app/typeit4me/id412141729?mt=12)

I think that I will switch to Espanso since it is open source and works with Linux, but there are a few features it still lacks that would be nice to have. Forgive me if I missed other threads that ask for some of these.

1. Support for pictures with embedded hyperlinks. There's probably a term for these but an example would be including the LinkedIn logo in your email signature that can be clicked to go to your LinkedIn page.
2. Support for formatted (rich) text, e.g. underline, bold, etc.
3. The option to enable notifications when you manually type text that is identical to (or better, similar to) text you've previously created a replacement for. This helps you remember to use replacements you've made in the past that you might have forgotten about.
4. A secure means of suggesting phrases that you type regularly that you haven't created a snippet for yet. TextExpander has something like this but it only suggests very short snippets that match each other exactly. I'd like to see something that could suggest long phrases that are similar in content or meaning but aren't necessarily identical. This would require some level of logging, hence the need for it to be secure. If the user could enable or disable it and control how many keys it could record at a time (encrypted, locally), that would be ideal.
5. Ability to use Espanso triggers inside Espanso forms. For example, I often want a date field where I can enter the date manually or run a trigger for a specific date format (but the trigger doesn't work):
![image](https://user-images.githubusercontent.com/46253001/223195126-403713ce-c597-4565-8b8e-fe9fbfd53d1b.png)

**Comment by ernstki** [comment]
> 1. Support for pictures with embedded hyperlinks. There's probably a term for these but an example would be including the LinkedIn logo in your email signature that can be clicked to go to your LinkedIn page.
> 2. Support for formatted (rich) text, e.g. underline, bold, etc.

These should already be working, you just have to know a little bit of HTML to write the match rule. See #214 for details.

I tested with a not-very-current release of Espanso on a Linux box with LibreOffice and [this web-based WYSIWYG editor](https://alex-d.github.io/Trumbowyg), and both of [Federico's examples](https://github.com/espanso/espanso/issues/214#issuecomment-713828898) as well as your proposed use case worked fine:

```yaml
matches:
  - trigger: ":sig"
    html: |
      <a href="https://espanso.org">
        <img style="display:block" src="https://espanso.org/assets/images/extended-logo-light-a8225538ce1dfbc807b6bd0529e84ea5.svg">
        Espanso: supercharge your typing experience!
      </a>
```

> 3. The option to enable notifications when you manually type text that is identical to (or better, similar to) text you've previously created a replacement for.
> 4. A secure means of suggesting phrases that you type regularly that you haven't created a snippet for yet.

Intriguing.

Also, stating a personal preference, I like low-noise tools, and Espanso in its current incarnation fits that bill. Not to minimize _your_ desire for these features, just to point out that "more Clippy" is not something _everyone_ wants.

> 5. Ability to use Espanso triggers inside Espanso forms. For example, I often want a date field where I can enter the date manually or run a trigger for a specific date format (but the trigger doesn't work):

I'll vote for this one! Seems like it should just _work_, but avoiding recursion is sure to be tricky…
